### PR TITLE
redeem and its test

### PR DESCRIPTION
**Pull request by cranium7811**

### CODE REVIEW

**Comment by 0xAlcibiades** [comment]
@cranium7811 can you resolve the merge conflicts here, please.

File: src/OptionSettlement.sol [hunk 1]
```diff
@@ -264,9 +264,50 @@ contract OptionSettlementEngine is ERC1155 {
...
         require(tokenType[claimId] == Type.Claim, "Token is not an claim");
-        // TODO(Implement)
+
+        Claim storage claimRecord = claim[claimId];
+
+        uint256 optionId = claimRecord.option;
+
+        Option storage optionRecord = option[optionId];
+
+        require(!claimRecord.claimed, "Already Claimed");
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:276]: This require could come before all the storage access on the option.

File: src/OptionSettlement.sol [hunk 2]
```diff
@@ -264,9 +264,50 @@ contract OptionSettlementEngine is ERC1155 {
...
+
+        uint256 balance = ERC1155(address(this)).balanceOf(msg.sender, claimId);
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:278]: You can just call this.balanceOf, no?

File: src/OptionSettlement.sol [hunk 3]
```diff
@@ -264,9 +264,50 @@ contract OptionSettlementEngine is ERC1155 {
...
+        require(balance == 1, "no claim token");
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:279]: This check can come before all the storage access.

File: src/OptionSettlement.sol [hunk 4]
```diff
@@ -264,9 +264,50 @@ contract OptionSettlementEngine is ERC1155 {
...
+
+        require(
+            optionRecord.expiryTimestamp <= block.timestamp,
+            "Not expired yet"
+        );
+
+        uint256 tx_amount = (optionRecord.exerciseAmount *
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:286]: These should be calculated as 2 entirely seperate amounts, and transfered conditionally. There may be either or both of these assets transferred on claim.

File: src/OptionSettlement.sol [hunk 5]
```diff
@@ -264,9 +264,50 @@ contract OptionSettlementEngine is ERC1155 {
...
+            claimRecord.amountExercised) +
+            (optionRecord.underlyingAmount *
+                (claimRecord.amountWritten - claimRecord.amountExercised));
+
+        SafeTransferLib.safeTransfer(
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:291]: Where is the conditional transfer for the exercise asset?

File: src/OptionSettlement.sol [hunk 6]
```diff
@@ -264,9 +264,50 @@ contract OptionSettlementEngine is ERC1155 {
...
+        uint256[] memory amounts = new uint256[](2);
+        amounts[0] = (claim[claimId].amountWritten -
+            claim[claimId].amountExercised);
+        amounts[1] = 1;
+
+        tokenType[claimId] = Type.None;
+        claimRecord.amountExercised = claimRecord.amountWritten;
+        claim[claimId].claimed = true;
+
+        _batchBurn(msg.sender, tokens, amounts);
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:310]: We don't need a batch burn here, we should only burn the claim.

File: src/OptionSettlement.sol [hunk 7]
```diff
@@ -264,9 +264,50 @@ contract OptionSettlementEngine is ERC1155 {
...
+            msg.sender,
+            tx_amount
+        );
+
+        uint256[] memory tokens = new uint256[](2);
+        tokens[0] = optionId;
+        tokens[1] = claimId;
+
+        uint256[] memory amounts = new uint256[](2);
+        amounts[0] = (claim[claimId].amountWritten -
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:302]: We shouldn't be touching the option token here per se, it doesn't do us any good, just results in another sstore.

**Review by 0xAlcibiades — CHANGES_REQUESTED** [review]

**Comment by cranium7811** [comment]
created a new PR

**Pull request closed without merging**